### ISSUE TITLE
Improve UI for configuring commissions on products

### ISSUE THREAD

**Issue by mindctrl** [easydigitaldownloads/EDD-Commissions#93]
We've talked about this in Slack, but I don't see an issue here to track it.

Right now configuring a product's commissions settings is not very user friendly. You have to enter comma-separated user IDs and then you have to enter comma-separated rates in the box below. We've had a fair amount of tickets about this confusion, and today had a request from a customer to improve it.

It would be great to see this improved with some kind of user selection dropdown with the rate configuration on the same row.

**Comment by pippinsplugins** [comment]
What about a repeatable table similar to variable prices and file downloads?

Simply add a row, pick a user, and set a rate for each recipient.

**Comment by mindctrl** [comment]
That would be great.

**Comment by pippinsplugins** [comment]
I'll see if I can work on this sometime after EDD 2.5

**Comment by ghost1227** [comment]
@pippinsplugins Here's what I've got so far on a reworking of this meta box. Haven't rewritten all the calculation logic yet, but loading/saving work, as does adding/removing rates. I'd suggest that this rewrite negates the need for a separate issue for #154.

<img width="787" alt="screen shot 2017-01-11 at 4 34 42 pm" src="https://cloud.githubusercontent.com/assets/36433/21869171/e3d7bb7a-d81b-11e6-96b8-63dc0735384d.png">

**Comment by ghost1227** [comment]
@mindctrl You might want to weigh in as well.

**Comment by pippinsplugins** [comment]
@ghost1227 That's looking good! Is it backwards compatible with previous data?

**Comment by ghost1227** [comment]
Yes. Right now, it doesn't have an actual upgrade routine... for testing purposes, I've got it running a pseudo-upgrade on first load of any given download, but that's obviously not ideal long run. Once the actual coding is done, I'll add an upgrade routine that will update existing data to the new format.

**Comment by mindctrl** [comment]
@ghost1227 looks nice!

**Comment by pippinsplugins** [comment]
@ghost1227 If possible please make it work with the existing data formats so that no upgrade routine is required.

**Comment by ghost1227** [comment]
@pippinsplugins The existing data format is a comma separated string (for both fields). We'd have to write a custom save handler to convert back and forth...

**Comment by pippinsplugins** [comment]
That's fine. It's easy to convert a comma separated string into an array.

Make a helper function that converts from and one that converts to.

**Comment by ghost1227** [comment]
@pippinsplugins Ok, I wrote a helper function but in hindsight, I'm not 100% confident I'm doing what you actually want... Right now I've got it always storing as a string, converting to and from string to array for display. Seems less than optimal. Is that where you wanted to go?

**Comment by pippinsplugins** [comment]
Yep that's what I wanted! Can you show me an updated PR?

**Comment by ghost1227** [comment]
@pippinsplugins There's the code so far. Seems to work!

**Comment by pippinsplugins** [comment]
Great, let's get some testing on it soon after core 2.7 is released!

**Comment by ghost1227** [comment]
I'll note that there is some temporary code in it at the moment. In issue #148 I reworked how Commissions loads styles, allowing it to load on pages other than the settings page. Until that change is merged, this issue has a temporary block to bootstrap loading assets allowing the new JS to load in the right place.

**Comment by ghost1227** [comment]
This also definitively resolves #154

**Comment by pippinsplugins** [comment]
@ghost1227 This is working beautifully!

A few todos:

1. Please add a tool tip to the rate field that explains that the user's rate and the global rate will be used if no rate is entered: 
![screen shot 2017-01-25 at 11 08 05 am](https://cloud.githubusercontent.com/assets/1034109/22300910/a6f6a3c8-e2ee-11e6-8989-d1ca881eb58d.png)

2. Let's add a tool tip to `Type` as well that better explains with examples the two types.

**Comment by ghost1227** [comment]
Reminder: Anyone who reviews this issue leave feedback on tooltip content.

**Comment by pippinsplugins** [comment]
See PR

**Comment by ghost1227** [comment]
Updated

**Comment by pippinsplugins** [comment]
:+1:

**Comment by mintplugins** [comment]
Just tested this. My user's drop down is currently not showing any users. I have cleared my cache and there are no errors in the js console. 

<img width="395" alt="screen shot 2017-02-08 at 12 48 34 pm" src="https://cloud.githubusercontent.com/assets/1638128/22749929/fd7788a2-edfc-11e6-8b8a-4bb9893797e0.png">

**Comment by pippinsplugins** [comment]
@ghost1227 could you take a look?

**Comment by cklosowski** [comment]
Just as a heads up, this problem is due to the 'Chosen' field being hidden
before it's shown, we show/hide the wrapper we just need to do a
`$(element).css('width', '100%')`, targeting the Chosen Wrapper.

**Comment by ghost1227** [comment]
I haven't been able to replicate the issue, but it sounds like @cklosowski is right. However, since that field is generated using the HTML elements class I'm not sure how to adjust it on the fly.

**Comment by cklosowski** [comment]
You have to do it when you click the 'enable commissions' for the download.

**Comment by ghost1227** [comment]
@cklosowski That doesn't work right either. If entries are already saved, they don't run through that, and any time you add a new row it reverts to the default width (which would be broken in his case).

**Comment by mintplugins** [comment]
Once the issue with chosen JS is working, just a quick note that we'll need to make sure the Simple Shipping integration still works. I tested it with Simple Shipping and the additional options are not currently being output.

**Comment by ghost1227** [comment]
@mintplugins can you elaborate?

**Comment by mintplugins** [comment]
When Simple Shipping is active, it outputs additional options into the commissions metabox to determine how Shipping is split. 
https://github.com/easydigitaldownloads/EDD-Commissions/blob/release/3.3/includes/integrations/simple-shipping.php

It only works if you are using percentage based commissions so there was some additional js to hide/show options accordingly as well. 
https://github.com/easydigitaldownloads/EDD-Commissions/blob/release/3.3/assets/js/admin-eddc-shipping-integration.js

**Comment by cklosowski** [comment]
As well as the Recurring addon that's in #113

**Comment by ghost1227** [comment]
@mintplugins The patch chris just pushed works for me... can you verify?

**Comment by mintplugins** [comment]
That fixes the width issue with the chosen field. 

Note that saving it appears to have no effect though now.

**Comment by ghost1227** [comment]
You're saying saving doesn't work at all now?

**Comment by mintplugins** [comment]
Just saving doesn't work. 

Before save:
<img width="1021" alt="screen shot 2017-02-10 at 2 30 40 pm" src="https://cloud.githubusercontent.com/assets/1638128/22840962/9498aada-ef9d-11e6-810b-2f067694e92e.png">

After save:
<img width="1020" alt="screen shot 2017-02-10 at 2 30 48 pm" src="https://cloud.githubusercontent.com/assets/1638128/22840973/9be175ec-ef9d-11e6-84c5-98d28ce2318d.png">

**Comment by mintplugins** [comment]
The original width issue is gone though.

**Comment by ghost1227** [comment]
@mintplugins ok, that was my bad. repull and test again

**Comment by ghost1227** [comment]
Also, the integration 'issue' isn't an issue. This PR doesn't touch the settings.php file, and the filter that is relevant to those integrations wasn't added until after this branch was created.

**Comment by mintplugins** [comment]
Getting closer. However, if you make a brand new product and add more than a single receiver, it seems to strip them out upon save.

**Comment by ghost1227** [comment]
![](http://i.giphy.com/SfYTJuxdAbsVW.gif)

... ok back to the drawing board

**Comment by mintplugins** [comment]
:joy:

**Comment by ghost1227** [comment]
@mintplugins Take 247, action!

**Comment by mintplugins** [comment]
Looks good now I think! 

Test 1:
- Tested making a new product on Master with commissions
- Switched to branch 93, updated, everything remains correct. 

Test 2:
- Tested a new post immediately adding multiple receivers and saved - all good. 

Test 3:
- Tested a new post immediately adding a single receiver and saved - all good.

**Comment by ghost1227** [comment]
@cklosowski @pippinsplugins Looks like this is ready for final review!

**Comment by mintplugins** [comment]
**Reminder to self:**
I still want to test actually generating some commissions in a few different scenarios in the above situations.

**Comment by mintplugins** [comment]
I found one issue. If you enable commissions but then don't actually enter a receiver, you will get an error upon checkout. The 3.3 branch allows the purchase to go through successfully without generating any commissions. Currently this fails with this error message:

```
Notice: Trying to get property of non-object in /Users/philjohnston/Documents/wps/edd2.dev/wp-content/plugins/EDD-Commissions/includes/commission-functions.php on line 805
```

**Comment by mintplugins** [comment]
However, that happens on master as well - which is where this branch was forked from. Upon merge into 3.3 that should resolve this. Will need to do a test after merge with commissions enabled but no receiver chosen.

**Comment by mintplugins** [comment]
@ghost1227 This is working perfectly now in all tests mentioned above and actually generating commissions after test purchases.

**Comment by ghost1227** [comment]
@pippinsplugins Ready for final approval

**Comment by pippinsplugins** [comment]
The issue with the select fields being broken isn't fixed for me. To replicate:

1. Collapse the metabox.
2. Reload the page
3. Open the metabox

**Comment by ghost1227** [comment]
Verified. Fails if the metabox itself is collapsed, works if the metabox is expanded in all cases. @cklosowski I'm going to review, but the code for this was yours... if you know a solution please let me know

**Comment by ghost1227** [comment]
@pippinsplugins That should fix it

**Comment by pippinsplugins** [comment]
@ghost1227 close. It works now when clicking on the actual arrow icon in the top right of the metabox, but fails when clicking on the metabox title, which also collapses / expands the box.

**Comment by ghost1227** [comment]
@pippinsplugins Good point! Fixed

**Comment by pippinsplugins** [comment]
Yep, there we go!

**Comment by pippinsplugins** [comment]
Small merge conflict, could you update that please @ghost1227 ?

**Comment by ghost1227** [comment]
@pippinsplugins Not sure how small 100% of a file is... but fixed!

**Comment by pippinsplugins** [comment]
Merged!

**Comment by ghost1227** [comment]
Awesome! Thanks!

**Comment by cklosowski** [comment]
Reopening this...something we need to do, the `label` needs to tied to the checkbox...so you can click the text to check the box:

![screen shot 2017-02-16 at 3 04 30 pm](https://cloud.githubusercontent.com/assets/1390124/23043377/50ed48e8-f459-11e6-8fb2-1cb8e4c4097e.png)

**Comment by cklosowski** [comment]
@ghost1227 can you fix up that label to have the `for` attribute for the `id` of the checkbox?

**Comment by cklosowski** [comment]
Ignore me...The I had previously tested this so the metabox was in the new spot, but the old UI...the new UI has the correct label behavior...
nothing to see here....
carry on.

**Comment by cklosowski** [comment]
Correction I found 2 issues:
1) If you only have 1 commission author and do not enter a rate, it doesn't save. (2 or more authors, or adding a rate saves)
2) The 'remove' icon presents the warning that you have to have at least one row when you have more than 1 row.

**Comment by ghost1227** [comment]
Fixed issue 1. Stupid logic error. Issue two is a conflict with the EDD core JS. It's triggering the repeatable row JS when you hit remove.

**Comment by cklosowski** [comment]
I'll look at the JS trigger

**Comment by cklosowski** [comment]
Ok give this a shot @mintplugins @ghost1227 

We needed to do some work on the 'remove repeatable' icon...which would have required more CSS...what do you guys think about just using Dashicons which is already required in WP Admin?

![screen shot 2017-02-16 at 5 14 25 pm](https://cloud.githubusercontent.com/assets/1390124/23047190/676d6208-f46b-11e6-9ee5-0082cdf33b52.png)

**Comment by mintplugins** [comment]
Testing now

**Comment by mintplugins** [comment]
While everything seems to be saving correctly, if I actually make a purchase while the rate is empty, it isn't falling back to the site default for me. Instead, a commission of 0% is generated - as if 0 was entered as the commission rate for the user. 

Also, potentially related, if I enter the same Vendor twice for the same product, their rates actually save as "0" instead of being empty:

Before save:
<img width="1011" alt="screen shot 2017-02-16 at 11 07 10 pm" src="https://cloud.githubusercontent.com/assets/1638128/23052108/b6c3d090-f49c-11e6-88aa-2bdf595d508c.png">

After save:
<img width="1027" alt="screen shot 2017-02-16 at 11 07 18 pm" src="https://cloud.githubusercontent.com/assets/1638128/23052111/bde5c824-f49c-11e6-872a-75a82eced8d0.png">

**Comment by ghost1227** [comment]
@mintplugins try it now. I screwed up the saving logic and was setting blank to zero.

@cklosowski Breaks continuity with the other meta boxes. More CSS or not, I think a uniform experience is better.

**Comment by cklosowski** [comment]
@ghost1227 That's a shame, I really dislike our current `x` system, since it's a weird CSS hover effect that is just not needed.

I'd like to try swtiching everythying over to this new way to delete. it does a few things for us:
1) Uses dashicons...so it's native
2) Uses no Inline CSS or CSS in our admin includes
3) Gives a larger 'hit' area on mobile, thank our `x` does.

**Comment by ghost1227** [comment]
@cklosowski I'm COMPLETELY ok with changing it, just think if we're going to change it, it should be done everywhere.

**Comment by cklosowski** [comment]
@ghost1227 and it will be, as we work through extensiosn and core.

**Comment by cklosowski** [comment]
@ghost1227 @mintplugins @SDavisMedia  if we're good with this new `dismiss` and things are working well, we good to merge this beast in?

**Comment by cklosowski** [comment]
For reference, the new `dismiss` icon for removing items:
![screen shot 2017-02-17 at 10 03 47 am](https://cloud.githubusercontent.com/assets/1390124/23085919/b44f475a-f527-11e6-9da6-5a694ab74e42.png)

**Comment by mintplugins** [comment]
Things look good with the dismiss button here. 

Can we give a unique identifying name to the "type" setting? Either a CSS class or ID would be sweet. Right now it is just "edd_field_type_select"

It would be good if it has a unique ID similar to how the eddc_commission_rates_wrapper does.

**Comment by mintplugins** [comment]
Actually the parent element to that is likely better to get the ID. 

The one called eddc_toggled_row

**Comment by mintplugins** [comment]
Also note that we need to make sure the Shipping and Recurring JS integrations work as well. They are now merged into this branch so they can be fixed here.

**Comment by ghost1227** [comment]
Latest change from @cklosowski gets a 👍 from me. As far as I'm concerned, this is good to go. Can we get approval from one more person?

**Comment by pippinsplugins** [comment]
@ghost1227 @cklosowski rates don't save for me anymore. No matter what I enter, no rate is saved. The user field also isn't saved.

**Comment by ghost1227** [comment]
@pippinsplugins Try again now, please!

**Comment by pippinsplugins** [comment]
Works great now.

**Comment by mintplugins** [comment]
Here's what I can think of that should be tested prior to merge:

- Visual Check: Make sure all Commissions settings hide if unchecked and show if checked. 

Purchase the following combinations:

- Free product with Single receiver and 50% rate. 
Result: 

- Single receiver and 50% rate. 
Result: 

- Single receiver and blank rate in product, 50% user rate. 
Result: 

- Single receiver, blank rate, blank user rate, site wide default of 50%. 
Result: 

- Single receiver, blank rate, blank user rate, blank site wide default rate. 

- 2 receivers 50% rate each. 
Result:

- 2 receivers, rates blank, user rates 50%. 
Result:

- 2 receivers, rates blank, user rates blank, site rate 50%.
Result:

- 2 receivers, first with 50% rate, second with blank rate and user rate at 50%. 
Result:

- 2 receivers, first with 50% rate, second with blank rate, blank user rate, 50% site default rate.
Result:

- 2 receivers, first with 50% rate, second with blank rate, blank user rate, and blank site default rate.
Result:

- 2 receivers, second with 50% rate, first with blank rate, 50% user rate. 
Result:

- 2 receivers, second with 50% rate, first with blank rate, blank user rate, and 50% site rate. 
Result:

- 2 receivers, second with 50% rate, first with blank rate, blank user rate, and blank site rate. 
Result:

**Shipping Integration test:**

- Visual Check: Make sure shipping only shows if Flat rate and Shipping are chosen. 
Result:

- Make sure the shipping options save. 
Result:

**Recurring Integration test:**

- Visual Check: Make sure recurring option only shows if Flat rate and Shipping are chosen. 
Result:

- Make sure the recurring option saves. 
Result:

**Comment by mintplugins** [comment]
Testing list:

- Visual Check: Make sure all Commissions settings hide if unchecked and show if checked. 
**Result**: Core Commissions settings do not hide for me unless I click whitespace after unchecking the "Enable Commissions" metabox. However, Shipping and Recurring integrations do hide immediately after unchecking. Tried to confirm with @cklosowski but he was unable to replicate. This could be a weird issue on my end only so we'll need someone else to confirm this. 

Purchase the following combinations:

- Free product with Single receiver and 50% rate. 
**Result**: Correct: No Commission Generated .

- Single receiver and 50% rate. 
**Result**: Correct: Proper commission generated. 

- Single receiver and blank rate in product, 50% user rate. 
**Result**: Correct: Proper commission generated.

- Single receiver, blank rate, blank user rate, site wide default of 50%. 
**Result**: Correct: Proper commissions generated.

- Single receiver, blank rate, blank user rate, blank site wide default rate. 
**Result**: Incorrect. $0 Commission Generated. Should have no commissions generated. 

- 2 receivers 50% rate each. 
**Result**: Correct. 

- 2 receivers, rates blank, user rates 50%. 
**Result**: Correct. 

- 2 receivers, rates blank, user rates blank, site rate 50%.
**Result**: Correct.

- 2 receivers, first with 50% rate, second with blank rate and user rate at 50%. 
**Result**: Correct.

- 2 receivers, first with 50% rate, second with blank rate, blank user rate, 50% site default rate.
**Result**: Correct.

- 2 receivers, first with 50% rate, second with blank rate, blank user rate, and blank site default rate.
**Result**: Incorrect. $0 Commission was generated for user with no rate. 

- 2 receivers, second with 50% rate, first with blank rate, 50% user rate. 
**Result**: Correct. 

- 2 receivers, second with 50% rate, first with blank rate, blank user rate, and 50% site rate. 
**Result**: Correct.

- 2 receivers, second with 50% rate, first with blank rate, blank user rate, and blank site rate. 
**Result**: Incorrect. $0 Commission generated for first user. 

**Shipping Integration test:**

- Visual Check: Make sure shipping only shows if Flat rate and Shipping are chosen. 
**Result**: Correct.

- Make sure the shipping options save. 
**Result**: Correct.

**Recurring Integration test:** 

- Make sure the recurring option saves. 
**Result**: Correct.

**Comment by ghost1227** [comment]
The enable commissions display issue isn't happening for me, I'm guessing some sort of caching bug on your end.

**Comment by ghost1227** [comment]
The one outstanding issue brought up by @mintplugins has been determined to be not related to this issue and is of minor enough consequence that we tagged it for 3.3.3 https://github.com/easydigitaldownloads/EDD-Commissions/issues/198. Given the testing, I now recommend merging and closing this issue.

**Comment by cklosowski** [comment]
Re-opening just to get some testing done once #178 is done.

**Comment by pippinsplugins** [comment]
The repeatable rows no longer work on #178. The add new button doesn't do anything.

**Comment by pippinsplugins** [comment]
Looks like the JS file isn't loaded anymore.

**Comment by SDavisMedia** [comment]
Confirmed. It's definitely not loading on `issue/178`.

**Comment by cklosowski** [comment]
Ok, should be fixed up.

**Comment by SDavisMedia** [comment]
Missing a space here: 
![0c7ab0547c40e57193cc65db547fa4f7 _image 202017-02-21 20at 208 22 49 20am](https://cloud.githubusercontent.com/assets/2359131/23173743/018a1722-f80f-11e6-9921-9cdf32eda828.png)

**Comment by mintplugins** [comment]
We should add "eddc_toggled_row" as a class for the Shipping and Recurring integrations instead of "eddc_commission_row".

**Comment by mintplugins** [comment]
Nevermind that :P Just spoke with @cklosowski who made me realize why they need their own class names.
